lib/saluki-components/src/destinations/datadog_metrics/endpoint/endpoints.rs

```diff
@@ -0,0 +1,152 @@
+        &self.api_keys
+    }
+
+    fn add_api_key(&mut self, api_key: String) {
+        self.api_keys.push(api_key);
+    }
+
+    fn set_domain(&mut self, domain: String) {
+        self.domain = domain;
+    }
+}
+
+/// Converts AdditionalEndpoints into a list of SingleDomainResolvers with its
+/// destination domain and api keys.
+pub fn create_single_domain_resolvers(
+    endpoints: &AdditionalEndpoints,
+) -> Result<Vec<SingleDomainResolver>, EndpointError> {
+    let mut resolvers = Vec::new();
+    for (domain, api_keys) in &endpoints.0 {
+        let new_domain = add_adp_version_to_domain(domain.to_string())?;
+        let mut seen = HashSet::new();
+        let mut resolver = SingleDomainResolver::default();
+        for api_key in api_keys {
+            let trimmed_api_key = api_key.trim();
+            if trimmed_api_key.is_empty() || seen.contains(trimmed_api_key) {
+                continue;
+            }
+            seen.insert(trimmed_api_key);
+            resolver.add_api_key(trimmed_api_key.to_string());
+        }
+        if !resolver.api_keys.is_empty() {
+            resolver.set_domain(new_domain.trim_end_matches("/").to_string());
+            resolvers.push(resolver);
+        }
+    }
+    Ok(resolvers)
+}
+
+impl FromStr for AdditionalEndpoints {
+    type Err = serde_json::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let endpoints: HashMap<String, Vec<String>> = serde_json::from_str(s)?;
+        Ok(AdditionalEndpoints(endpoints))
+    }
+}
+
+/// Return the correct prefix for a domain.
+pub fn get_domain_prefix(app: String) -> String {
+    let app_details = saluki_metadata::get_app_details();
+    let version = app_details.version();
+    format!(
+        "{}-{}-{}-{}.agent",
+        version.major(),
+        version.minor(),
+        version.patch(),
+        app
+    )
+}
+
+/// Prefixes the domain with the ADP version.
+pub fn add_adp_version_to_domain(domain: String) -> Result<String, EndpointError> {
+    let dd_url_regexp = Regex::new(r"^app(\.mrf)?(\.[a-z]{2}\d)?\.(datad(oghq|0g)\.(com|eu)|ddog-gov\.com)$").unwrap();
+
+    let mut new_domain = domain.clone();
+
+    // Url::Parse doesnt work without this
+    if !domain.starts_with("https://") && !domain.starts_with("http://") {
+        new_domain = format!("https://{}", domain);
+    }
+
+    let d = new_domain.clone();
+    let url = Url::parse(&new_domain).context(Parse { domain })?;
+
+    if let Some(host) = url.host_str() {
+        // Do not update unknown URLs
+        if !dd_url_regexp.is_match(host) {
+            return Ok(host.to_string());
+        }
+        let subdomain = host.split('.').next().unwrap_or("");
+
+        let new_subdomain = get_domain_prefix("adp".to_string());
+
+        let new_host = &host.replacen(subdomain, &new_subdomain, 1);
+
+        return Ok(new_host.to_string());
+    }
+    Err(EndpointError::Parse {
+        source: url::ParseError::EmptyHost,
+        domain: d,
+    })
```

Comment:
You can make this a little more idiomatic (and some of this also is meant to work in tandem with the above suggestion) by using a `match` block:

```suggestion
    match url.host_str() {
        Some(host) => {
            // Do not update unknown URLs
            if !dd_url_regexp.is_match(host) {
                return Ok(host.to_string());
            }

            let subdomain = host.split('.').next().unwrap_or("");
            let new_subdomain = get_domain_prefix("adp");
            let new_host = host.replacen(subdomain, &new_subdomain, 1);

            Ok(new_host)
        },
        None => Err(EndpointError::Parse {
            source: url::ParseError::EmptyHost,
            domain: new_domain,
        }),
    }
```

(I also removed some stuff like the reference on `&host.replacen(...)` and then the `new_host.to_string()` because you already get an owned string back from `replacen`)